modules/virt-about-application-consistent-backups.adoc

```diff
@@ -0,0 +1,13 @@
+// Module included in the following assemblies:
+//
+// * virt/backup_restore/virt-backup-restore-snapshots.adoc
+
+:_mod-docs-content-type: CONCEPT
+[id="virt-about-application-consistent-backups_{context}"]
+= About application-consistent snapshots and backups
+
+You can configure application-consistent snapshots and backups for Linux or Windows virtual machines (VMs) through a cycle of freezing and thawing. For any application, you can either configure a script on a Linux VM or register on a Windows VM to be notified when a snapshot or backup is due to begin.
+
+On a Linux VM, freeze and thaw processes trigger automatically when a snapshot is taken or a backup is started by using, for example, a plugin from Velero or another backup vendor. The freeze process, performed by QEMU Guest Agent (QEMU GA) freeze hooks, ensures that before the snapshot or backup of a VM occurs, all of the VM's filesystems are frozen and each appropriately configured application is informed that a snapshot or backup is about to start. This notification affords the applications the opportunity to quiesce the filesystems before the guest agent invokes a main hook script, which then calls other application-specific scripts that temporarily deactivate, or freeze, all guest VM applications. All pending application operations are flushed, or stopped, and all new connection requests are refused. When all applications have become inactive, the QEMU GA freezes the filesystems, and a snapshot is taken or a backup initiated. After the taking of the snapshot or start of the backup, the thawing process begins. The filesystems begin to accept write requests, and individual hook scripts (via a main hook script) then resume or restart applications that had been inactivated during the freeze process.
```

Comment:
I would correct this just a little bit:
- During a freeze operation, applications are first notified and have the opportunity to quiesce their state.  This may be different for different applications and may involve: temporarily refusing new requests, finishing in-progress operations, and flushing data to disk.
- Next, the operating system is directed to quiesce the filesystems by flushing outstanding writes to disk and freezing new write activity
- During thaw, filesystem writing is reactivated and applications are notified to resume normal operations.

Your wording is good here but the applications do not quiesce the filesystems (they do flush their own caches).  Fs freeze and thaw is handled by a call to the operating system and should happen after application scripts have finished.  Do you know if the RHEL docs have any information about how to configure a Linux application for this mechanism?  I haven't been able to find a great resource suitable for linking here.

Reply:
I will do some looking and if I find nothing I will ask one of the RHEL writers. Thanks!

Reply:
That seems like an important gap to fill.  I the meantime I am working with Bill Bell and another interested engineer on a blog/cheatsheet to provide an end-to-end example of deploying a custom script to be used in this flow.  I will update you as that progresses.